<a href="https://colab.research.google.com/github/LOVELYVD/Feature-Engineering/blob/main/Feature_Engineering_One_Hot_Encoding_For_Multi_Categorical_Variables.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **One Hot Encoding For Multi Categorical Variables**

kaggle dataset:mercedsbenz.csv

**load categorical variables**

In [1]:
import numpy as np
import pandas as pd

#load categorical variables of dataset

data=pd.read_csv('mercedesbenz.csv',usecols=['X1','X2','X3','X4','X5','X6'])
data.head()

X1  X2 X3 X4 X5 X6
0  v  at  a  d  u  j
1  t  av  e  d  y  l
2  w   n  c  d  x  j
3  t   n  f  d  x  l
4  v   n  f  d  h  d

**How many unique categories in each and every column**

In [2]:
#unique label in each column variables
for col in data.columns:
  print(col,': ',len(data[col].unique()),'labels')

X1 :  27 labels
X2 :  44 labels
X3 :  7 labels
X4 :  4 labels
X5 :  29 labels
X6 :  12 labels


**Lets check how many columns will be generated after one hot encoding to 6 variables**

In [3]:
pd.get_dummies(data,drop_first=True).shape

(4209, 117)

just from 6 initial categorical variables we generate 117 new variables


As number of columns/variables increase,it leads to curse of dimensionality whichh will impact acuracy


**SOLUTION**:

Take **most frequent** 10 categories/**label**s from each features nad skip the rest from each column.

Apply one hot encoding to those 10 frequent labels of the variable.

1 Binary variable for each of 10 frequent labels of the variable only.

Group all the rest under a new category and drop 

Finding topmost 20 values/categories of  X2 variable

In [4]:
 data.X2.value_counts().sort_values(ascending=False).head(20)

as    1659
ae     496
ai     415
m      367
ak     265
r      153
n      137
s       94
f       87
e       81
aq      63
ay      54
a       47
t       29
i       25
k       25
b       21
ao      20
z       19
ag      19
Name: X2, dtype: int64

Lets take top 10  frequent categories of the variable X2 from previous step as a list

In [9]:
top_10_labels=[x for x in data.X2.value_counts().sort_values(ascending=False).head(10).index]
# index edukumbol we get the categories name
top_10_labels

['as', 'ae', 'ai', 'm', 'ak', 'r', 'n', 's', 'f', 'e']

Next we will make binary variables

ensemble technique

In [10]:
for label in top_10_labels:
  data[label]=np.where(data['X2']==label,1,0)
  #label indel replace 1 else 0
data[['X2']+top_10_labels].head(40)
#idhupole create matrix for all categories

X2  as  ae  ai  m  ak  r  n  s  f  e
0   at   0   0   0  0   0  0  0  0  0  0
1   av   0   0   0  0   0  0  0  0  0  0
2    n   0   0   0  0   0  0  1  0  0  0
3    n   0   0   0  0   0  0  1  0  0  0
4    n   0   0   0  0   0  0  1  0  0  0
5    e   0   0   0  0   0  0  0  0  0  1
6    e   0   0   0  0   0  0  0  0  0  1
7   as   1   0   0  0   0  0  0  0  0  0
8   as   1   0   0  0   0  0  0  0  0  0
9   aq   0   0   0  0   0  0  0  0  0  0
10   r   0   0   0  0   0  1  0  0  0  0
11   e   0   0   0  0   0  0  0  0  0  1
12  ai   0   0   1  0   0  0  0  0  0  0
13   e   0   0   0  0   0  0  0  0  0  1
14  as   1   0   0  0   0  0  0  0  0  0
15  ak   0   0   0  0   1  0  0  0  0  0
16   e   0   0   0  0   0  0  0  0  0  1
17  as   1   0   0  0   0  0  0  0  0  0
18   m   0   0   0  1   0  0  0  0  0  0
19  ak   0   0   0  0   1  0  0  0  0  0
20   m   0   0   0  1   0  0  0  0  0  0
21  as   1   0   0  0   0  0  0  0  0  0
22   a   0   0   0  0   0  0  0  0  0  0
23   m   0   0   0  1   0  0  0  0  0  0
24  ak   0   0   0  0   1  0  0  0  0  0
25   m   0   0   0  1   0  0  0  0  0  0
26  as   1   0   0  0   0  0  0  0  0  0
27  as   1   0   0  0   0  0  0  0  0  0
28  as   1   0   0  0   0  0  0  0  0  0
29   m   0   0   0  1   0  0  0  0  0  0
30  as   1   0   0  0   0  0  0  0  0  0
31  as   1   0   0  0   0  0  0  0  0  0
32   m   0   0   0  1   0  0  0  0  0  0
33  as   1   0   0  0   0  0  0  0  0  0
34  as   1   0   0  0   0  0  0  0  0  0
35  ai   0   0   1  0   0  0  0  0  0  0
36  as   1   0   0  0   0  0  0  0  0  0
37   k   0   0   0  0   0  0  0  0  0  0
38  as   1   0   0  0   0  0  0  0  0  0
39  as   1   0   0  0   0  0  0  0  0  0

In [11]:
# get whole set of dummy variables, for all the categorical variables

def one_hot_encoding_top_x(df, variable, top_x_labels):
  #variable#column name,
    # function to create the dummy variables for the most frequent labels
    # we can vary the number of most frequent labels that we encode
    
    for label in top_x_labels:
        df[variable+'_'+label] = np.where(df[variable]==label, 1, 0)

In [12]:
# read the data again
df = pd.read_csv('mercedesbenz.csv', usecols=['X1', 'X2'])

# encode X2 into the 10 most frequent categories
one_hot_encoding_top_x(df, 'X2', top_10_labels)
df.head()

X1  X2  X2_as  X2_ae  X2_ai  X2_m  X2_ak  X2_r  X2_n  X2_s  X2_f  X2_e
0  v  at      0      0      0     0      0     0     0     0     0     0
1  t  av      0      0      0     0      0     0     0     0     0     0
2  w   n      0      0      0     0      0     0     1     0     0     0
3  t   n      0      0      0     0      0     0     1     0     0     0
4  v   n      0      0      0     0      0     0     1     0     0     0

# **Advantages of OHE**



1.   Straightforward ti implement and doesnot require hours of variable exploration
2.   Doesnot expand massively the feature space



# **Disadvantages of OHE**


1.   Doesnot add info that may make the variable more predictive
2.   Doesnot keep info of ignored labels



So here we consider dominating categorical variables and considered the minorities as noise